src/gui/qgsmaptoolcapture.cpp

```diff
@@ -347,18 +347,42 @@ int QgsMapToolCapture::nextPoint( const QPoint &p, QgsPoint &layerPoint, QgsPoin
 
 int QgsMapToolCapture::addVertex( const QgsPoint& point )
 {
+  return addVertex( point, QgsPointLocator::Match() );
+}
+
+int QgsMapToolCapture::addVertex(const QgsPoint& point, QgsPointLocator::Match match )
+{
   if ( mode() == CaptureNone )
   {
     QgsDebugMsg( "invalid capture mode" );
     return 2;
   }
 
-  int res;
   QgsPoint layerPoint;
-  res = nextPoint( point, layerPoint );
-  if ( res != 0 )
+  QgsVectorLayer* vlayer = qobject_cast<QgsVectorLayer *>( mCanvas->currentLayer() );
+  if ( match.isValid() && match.hasVertex() &&
+       ( match.layer()->crs() == vlayer->crs() ) )
```

Comment:
When snapping to intersections is turned on, a match to an intersection is "vertex" match with no layer associated - we need to check for nullptr in match.layer()